File: frontend/src/types/player.ts
export type EloTier = 'milk' | 'beer' | 'highball' | 'wine' | 'soju' | 'whisky' | 'vodka'
export type DominantHand = 'left' | 'right'

export interface Player {
  id: string
  nickname: string
  avatar_url: string | null
  elo_score: number
  elo_level: EloTier
  line_id: string | null
  dominant_hand: DominantHand | null
  tiktok: string | null
  instagram: string | null
  is_active: boolean
  created_at: string
  member_seq: number
  member_code: string
  games: number
  wins: number
  draws: number
  losses: number
}

/** Mirrors backend app/models/player.py PlayerStats. */
export interface PlayerStats {
  player: Player
  games: number
  wins: number
  draws: number
  losses: number
  points: number
  avg_points: number
  score_percent: number
}

/** The opponent this player has faced most often ("เทกันจัง"). */
export interface NemesisInfo {
  player: Player
  encounters: number
  wins: number
  losses: number
  draws: number
}

/** Mirrors backend app/models/player.py PlayerProfile. */
export interface PlayerProfile {
  player: Player
  games: number
  wins: number
  draws: number
  losses: number
  points: number
  avg_points: number
  score_percent: number
  nemesis: NemesisInfo | null
  elo_rank: number
  total_ranked_players: number
  similar_players: Player[]
}

File: backend/app/models/player.py
from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, computed_field

EloTier = Literal["milk", "soju", "beer", "whisky", "highball", "vodka"]
DominantHand = Literal["left", "right"]


class PlayerBase(BaseModel):
    nickname: str
    line_id: str | None = None
    dominant_hand: DominantHand | None = None
    tiktok: str | None = None
    instagram: str | None = None


class PlayerCreate(PlayerBase):
    elo_score: int | None = None
    """Starting rating for a player who isn't a true beginner — defaults to
    elo_service.STARTING_SCORE (1000) when omitted."""


class PlayerUpdate(BaseModel):
    nickname: str | None = None
    line_id: str | None = None
    avatar_url: str | None = None
    is_active: bool | None = None
    dominant_hand: DominantHand | None = None
    tiktok: str | None = None
    instagram: str | None = None


class Player(PlayerBase):
    id: UUID
    avatar_url: str | None = None
    elo_score: int
    elo_level: EloTier
    is_active: bool
    created_at: datetime
    member_seq: int

    @computed_field  # type: ignore[prop-decorator]
    @property
    def member_code(self) -> str:
        """Display code e.g. "ND007" — derived from member_seq, not stored."""
        return f"ND{self.member_seq:03d}"


class PlayerStats(BaseModel):
    """Aggregated per-player stats for the Member List / Ranking views."""

    player: Player
    games: int
    wins: int
    draws: int
    losses: int
    points: int
    avg_points: float
    score_percent: float


class NemesisInfo(BaseModel):
    """The opponent this player has faced most often ("เทกันจัง")."""

    player: Player
    encounters: int
    wins: int
    losses: int
    draws: int


class PlayerProfile(BaseModel):
    """Single-player detail page — Member List stats plus their nemesis."""

    player: Player
    games: int
    wins: int
    draws: int
    losses: int
    points: int
    avg_points: float
    score_percent: float
    nemesis: NemesisInfo | None = None
    elo_rank: int
    total_ranked_players: int
    similar_players: list[Player] = []
